### ISSUE TITLE
Authentication with KeytabCredential fails with null password

### ISSUE THREAD

**Issue by scobb** [SteveSyfuhs/Kerberos.NET#266]
**Describe the bug**
First--I want to caveat that it's possible this is user error, and that I have not correctly configured the client or credential.

Motivation: I want to accept keytabs as  base64-encoded strings and validate in C# that they are valid to Authenticate against an AD server.

Background: I am running an AD server in windows server. I used `ktpass` to generate a keytab for a user and powershell to convert to a base64 string. 

The bug: I construct a `KeytabCredential` and attempt to authenticate, but observe a `System.AggregateException` with `Kerberos.NET.KerberosProtocolException: KDC KDC_ERR_PREAUTH_FAILED: Pre-authentication information was invalid` as inner exception.

Upon inspecting the `KeytabCredential`, I found the password is empty, which is consistent with the Preauth failing.

I can `Authenticate` with this user using a `KerberosPasswordCredential`.

**To Reproduce**
```
var ktBase64String = "myBase64Keytab==";
var domain = "my.domain";
var principal = "principal@my.domain";
var keytab = new KeyTable(Convert.FromBase64String(ktBase64String));


var cred = new KeytabCredential(principal, keytab, domain);
client.Authenticate(cred).Wait(); // exception

// this works - but we are trying to validate the keytab prior to passing it to linux container
var password = "a-very-strong-password";
var cred2 = new KerberosPasswordCredential(principal, password, domain);
client.Authenticate(cred2).Wait(); // tada!

```

**Expected behavior**
I would expect the keytab credential to be preserved in a manner that the AD server could authenticate.

**Screenshots**
n/a

**Additional context**
Full stack trace:
```
Unhandled Exception: System.AggregateException: One or more errors occurred. ---> Kerberos.NET.KerberosProtocolException: KDC KDC_ERR_PREAUTH_FAILED: Pre-authentication information was invalid
   at Kerberos.NET.Transport.KerberosTransportBase.Decode[T](ReadOnlyMemory`1 response) in D:\a\1\s\Kerberos.NET\Client\Transport\KerberosTransportBase.cs:line 70
   at Kerberos.NET.Transport.TcpKerberosTransport.<ReadResponse>d__13`1.MoveNext() in D:\a\1\s\Kerberos.NET\Client\Transport\TcpKerberosTransport.cs:line 133
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Kerberos.NET.Transport.TcpKerberosTransport.<SendMessage>d__11`1.MoveNext() in D:\a\1\s\Kerberos.NET\Client\Transport\TcpKerberosTransport.cs:line 61
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Kerberos.NET.Transport.KerberosTransportSelector.<SendMessage>d__6`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Kerberos.NET.Client.KerberosClient.<RequestTgt>d__90.MoveNext() in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 1059
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Kerberos.NET.Client.KerberosClient.<AuthenticateCredential>d__71.MoveNext() in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 391
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Kerberos.NET.Client.KerberosClient.<Authenticate>d__70.MoveNext() in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 338
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Setup.CustomAction.Program.Main(String[] args) in ...
```

**Comment by SteveSyfuhs** [comment]
Yeah, that's an interesting issue.

Can you list the contents of the keytab, minus the keys? If you're unsure how you can use the [Bruce](https://github.com/dotnet/Kerberos.NET#ktpass) tool with the command `ktpass --file c:\foo\blah.keytab`.

Looking over the source I see something kinda interesting that might be affecting it if you're missing an RC4 key: https://github.com/dotnet/Kerberos.NET/blob/develop/Kerberos.NET/Credentials/KeytabCredential.cs#L47-L6

It skips over it if it's doing an optimistic auth, which maybe isn't right. I'll have to look at it.

Also, you might try adding this bit of code before calling `Authenticate(cred)`

```csharp
client.Configuration.Defaults.AllowWeakCrypto = true;
```

If this line of code fixes it then it's definitely a bug that needs to be fixed. It shouldn't be requiring RC4. If that doesn't fix it then I'll need to step through it a bit more.

**Comment by scobb** [comment]
Thanks for the quick response. 

Adding `AllowWeakCrypto` flag did not allow call to succeed--same trace as above.

I generated this keytab using `/crypto All` flag on windows server, which seemed to keys with create 5 encryption methods:
```
PS > ktpass /in test.tab
Existing keytab:

Keytab version: 0x502
keysize 43 principal@my.domain ptype 1 (KRB5_NT_PRINCIPAL) vno 7 etype 0x1 (DES-CBC-CRC) keylength 8 (...)
keysize 43 principal@my.domain ptype 1 (KRB5_NT_PRINCIPAL) vno 7 etype 0x3 (DES-CBC-MD5) keylength 8 (...)
keysize 51 principal@my.domain ptype 1 (KRB5_NT_PRINCIPAL) vno 7 etype 0x17 (RC4-HMAC) keylength 16 (...)
keysize 67 principal@my.domain ptype 1 (KRB5_NT_PRINCIPAL) vno 7 etype 0x12 (AES256-SHA1) keylength 32 (...)
keysize 51 principal@my.domain ptype 1 (KRB5_NT_PRINCIPAL) vno 7 etype 0x11 (AES128-SHA1) keylength 16 (...)
```

I'm using native powershell `ktpass` .. let me know if the Bruce tool provides more info that you need.

**Comment by SteveSyfuhs** [comment]
Hmm okay, that all looks right then. This kinda makes me think maybe the key is genuinely bad?

When you enable `AllowWeakCrypto` it tells the system that it can use RC4. The snippet I linked to above says if you're able to authenticate optimistically then do it (otherwise you need metadata from the server), and at worst it will use RC4. RC4 doesn't use salts so it relies solely on the raw password itself, which makes it super useful for debugging, albeit insecure for production. Since RC4 failed that suggests it might just be a bad key.

How are you generating the file with ktpass? What parameters are you passing? There's a somewhat annoying bug in the 2012 versions of the tool that if you tell it to prompt for password it'll incorrectly include the \n of the enter key in the password before generating the key. Passing the password through the command line doesn't have this issue.

If that's not the case would you be able to collect a network trace of this? Can you also run ktpass with the `+dumpsalt` parameter and include that? The salt _should_ match what's being returned in the AS-REQ error response.

**Comment by scobb** [comment]
OK, I have some good news here. This was indeed user error--I iterated using `kinit` tool directly to discover this failed login was not specific to `Kerberos.NET`, but related to my keytab. Specifically, I had generated the keytab with lowercase realm name.

Once I got a successful login with `kinit`, I was able to use the same keytab to successfully log in with `Kerberos.NET`.

Thank you for your patience and help here.
